toolchains/solidity/linter/core/src/server.rs

```diff
@@ -0,0 +1,135 @@
+        let diags_res = linter.parse_content(filepath, &params.text_document.text);
+
+        if let Ok(diags) = diags_res {
+            let diags = diags.iter().map(|d| diagnostic_from_lintdiag(d.clone())).collect();
+            eprintln!("diags: {:?}", diags);
+            self.client.as_ref().borrow().log_message(MessageType::INFO, "diags: ");
+            self.client.as_ref().borrow().publish_diagnostics(params.text_document.uri.clone(), diags, None);
+            //self.client.publish_diagnostics(params.text_document.uri, diags, None).await;
+        } else if let Err(e) = diags_res {
+            self.client.as_ref().borrow().log_message(MessageType::ERROR, e.to_string());
+        }
+        
+    }
+
+    fn did_change(&self, params: DidChangeTextDocumentParams) {
+        eprintln!("file changed!");
+        self.client.as_ref().borrow()
+            .log_message(MessageType::INFO, "file changed!");
+
+        let filepath = filepath_from_uri(&params.text_document.uri);
+        let mut linter = self.linter.borrow_mut();
+        let linter = linter.as_mut().unwrap();
+        let diags_res = linter.parse_content(filepath, &params.content_changes[0].text);
+
+        if let Ok(diags) = diags_res {
+            let diags = diags.iter().map(|d| diagnostic_from_lintdiag(d.clone())).collect();
+            eprintln!("diags: {:?}", diags);
+            self.client.as_ref().borrow().publish_diagnostics(params.text_document.uri.clone(), diags, None);
+            //self.client.publish_diagnostics(params.text_document.uri, diags, None).await;
+        } else if let Err(e) = diags_res {
+            self.client.as_ref().borrow().log_message(MessageType::ERROR, e.to_string());
+        }
+    }
+}
+
+pub fn filepath_from_uri(uri: &Url) -> String {
+    let path = uri.path();
+//    let path = path.strip_prefix("/").unwrap_or(path);
+    path.to_string()
+}
+
+fn diagnostic_from_lintdiag(diag: LintDiag) -> Diagnostic {
+    eprintln!("diag: {:?}", diag);
+    Diagnostic {
+        range: Range {
+            start: Position {
+                line: diag.range.start.line as u32 - 1,
+                character: diag.range.start.character as u32,
+            },
+            end: Position {
+                line: diag.range.end.line as u32 - 1,
+                character: diag.range.end.character as u32,
+            },
+        },
+        severity: Some(DiagnosticSeverity::WARNING),
+        code: None,
+        code_description: None,
+        source: Some("solidity-linter".to_string()),
+        message: diag.message,
+        related_information: None,
+        tags: None,
+        data: None,
+    }
+}
+
+pub fn run_server(args: Args) {
+    eprintln!("starting LSP server");
+    let server = LspStdioServer::new();
+    let _ = LspStdioServer::serve(server, |client| Backend { 
+        client,
+        config_file_path: args.rules_file,
+        _working_dir: args.project_path[0].clone(),
+        linter: RefCell::new(None)
+    });
+}
```

Comment:
refactor to not use args struct as it should only be used my main binary and instead just take config path and project_path as args